modules/home-environment.nix

```diff
@@ -216,24 +191,24 @@ in
 
   config = {
     assertions = [
-      (let
-        badFiles =
-          filter (f: hasPrefix "." (baseNameOf f))
-          (map (v: toString v.source)
-          (attrValues cfg.file));
-        badFilesStr = toString badFiles;
-      in
-        {
-          assertion = badFiles == [];
-          message = "Source file names must not start with '.': ${badFilesStr}";
-        })
+      {
+        assertion = config.home.username != "";
+        message = "Username could not be determined";
+      }
+      {
+        assertion = config.home.homeDirectory != "";
+        message = "Home directory could not be determined";
+      }
     ];
 
     home.sessionVariables =
       let
         maybeSet = name: value:
           listToAttrs (optional (value != null) { inherit name value; });
-      in
+      in {
+        USER = config.home.username;
+        HOME = config.home.homeDirectory;
```

Comment:
I'm not quite certain what is the use case for these? I wouldn't expect, e.g., `${HOME}` to expand to the value `config.home.homeDirectory` unless you use `with`. For example, putting

```
home.sessionVariables = {
  FOO = "${HOME}/foo";
};
```
in my configuration gives the error "undefined variable ‘HOME’". It works if I change to
```
home.sessionVariables = with config.home.sessionVariables; {
  FOO = "${HOME}/foo";
};
```
but compared to that I think
```
home.sessionVariables = {
  FOO = "${config.home.homeDirectory}/foo";
};
```
is not too bad…

Reply:
Sorry, I just noted that you already had the `with` in your initial comment and I kind of see the point of it. Let me ponder for a while :-)

Reply:
Ok, have pondered now and haven't been able to think of a use case that requires the injection of HOME and USER into `sessionVariables`. In all cases I can think of it should be OK to use HOME and USER directly, for example,
```
home.sessionVariables = rec {
  FOO = "$HOME/foo";
  BAR = "${FOO}/bar
};
```
The `${HOME}` and `${USER}` syntax does seem appealing to me because of the similarity with the other variable interpolation; but ultimately I don't think this reason enough for the added code complexity and pedagogic problem of "`home.sessionVariables` holds your session variables except for these special ones" :-)

It's quite possible that I'm missing some compelling use case, however. So I'm open for dissenting opinions.

Reply:
I mainly want this so one can use HOME in a `home.file` option, also the combination with `XDG_CONFIG_DIR`:
```
home.sessionVariables.XDG_CONFIG_DIR = "${HOME}/.config";

home.file."${if useXDG then "${XDG_CONFIG_DIR}/foo/config" else "${HOME}/.foorc"}".text = ...
```

Reply:
Hmm, good point. Looking at this and thinking about #95 I came up with this: https://github.com/rycee/home-manager/issues/95#issuecomment-336670329

Given my comment there, I would like to see what you think about changing the above to

```
home.file."${if useXDG then "${config.xdg.configDir}/foo/config" else ".foorc"}".text = …;
```

or

```
home.file."${if config.xdg.preferXdg then "${config.xdg.configDir}/foo/config" else ".foorc"}".text = …;
```

or (if we decide to always prefer XDG directories) perhaps even

```
xdg.configFile."foo/config".text = …;
```

Actually, could still support `config.xdg.preferXdg` by having

```
xdg.configFile."foo/config" = {
  fallbackTarget = ".foorc";
  text = …;
};
```

where `fallbackTarget` is used if `!config.xdg.preferXdg`.

Reply:
Thinking about it again, I have to agree with you, setting `HOME` and `USER` in `sessionVariables` doesn't have much purpose if you can just use `homeDirectory` and `username`. And these `xdg.*` options look nice, better than using env vars for it.

I'll remove the `HOME` and `USER` session variables, which will take away the ugly parts of this PR :)